### ISSUE TITLE
Fix "Can't perform a React state update on an unmounted component." console warning

### ISSUE THREAD

**Issue by garanico** [luceroweb/guess-the-movie#88]
**Date Seen**
3/17/22 12:17pm PST

**Versions**
Linux
Chrome
[develop branch](https://github.com/luceroweb/guess-the-movie/tree/develop)

**Bug Description**
Below is a screenshot of the error that shows in the console during game play. See below for steps to reproduce and behavior.
![bug-react-state-update](https://user-images.githubusercontent.com/82012415/158879621-37c84d01-afff-426f-8624-0b4573e85037.png)

**Steps to Reproduce**
1. Open the console prior to starting the game so that the console is available throughout gameplay.
2. Start the game
3. Answer a question correctly
4. Click  "Next Question" button to show a new question. 
5. As you are playing, the error will eventually show up in the console.

**Actual Behavior**
The screenshot above is the warning that shows up in the console during game play.  There does not seem to be a pattern as to when this error shows up.  Sometimes the warning shows immediately after a correct answer as been selected, other times you may not see the warning show for a long time.  The error does not appear to show when answering one of the hard-coded false questions.

**Expected Behavior**
There should be no red warnings in the console.

**Troubleshooting/Testing Steps Attempted**
None yet.

**Workaround**
None yet.

**Comment by Hiwot-Beshe** [comment]
**[Issues]**

The bug appears only when the multiple-choice question is played. When the multiple-choice page is correct, the warning appears only once when the next question button is pressed on the Correct Answer page for the first time.

**[solve]**

Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

**Here's what we tried: we took it out of useEffect from Multipchoices.**
```
const randomizeAnswer = (multipleAnswer) => {
   let currentIndex = multipleAnswer;
   while (currentIndex != 0) {
     const randomIndex = Math.floor(Math.random() * currentIndex);
     currentIndex--;
     [multipleAnswer[currentIndex], multipleAnswer[randomIndex]] = [
       multipleAnswer[randomIndex],
       multipleAnswer[currentIndex],
     ];
   }
 };
```


**In this way, we also used useRef.**

```
const unmounted = useRef (false)
 useEffect(() => {
   setCorrectAnswer("");
   if (!unmounted.current) {
     setCorrectAnswer(multipleAnswer[0]);
   }
   // if (runRandom) {
   //   randomizeAnswer(multipleAnswer);
   // }
   return () => {
     unmounted.current=true
   }
 }, []);
```
```
``useEffect(() => {
  let isMounted =true
  if(isMounted) {
  setCorrectAnswer(multipleAnswer[0]);
  if (runRandom) {
    randomizeAnswer(multipleAnswer);
  }
}
return () => {
  isMounted = false
}
}, [multipleAnswer]);
```
**References were taken from Google search results**

You can declare let isMounted = true inside useEffect,
which will be changed in the cleanup callback, as soon as the component is unmounted.
Before state updates, you now check this variable conditionally:

```
useEffect(() => {
  let isMounted = true;               // note mutable flag
  someAsyncOperation().then(data => {
    if (isMounted) setState(data);    // add conditional check
  })
  return () => { isMounted = false }; // cleanup toggles value, if unmounted
}, []);    
```     

In conclusion, we were unable to resolve the issue.

**Comment by alberme** [comment]
committed fixes for this issue

[fix(selectedMovie): reset selectedMovie on wrong answer](https://github.com/luceroweb/guess-the-movie/commit/5940558deb7d6e6686dcbf2f6180b364c75500ca)
[fix(state update): add condition for empty selectedMovie](https://github.com/luceroweb/guess-the-movie/commit/b26188f60eae21eaeef857ac2f1ac0419b7944d3)
[chore(merge): merge from stash to this branch](https://github.com/luceroweb/guess-the-movie/commit/455a4203b3b46a599f9c6ada566ea42f1e73ae37)
